tests/README.md

```diff
@@ -0,0 +1,120 @@
+
+To create a new network, the `NetBuilder` should be used:
+
+```rust
+// Create a network of 10 nodes, out of which 3 are faulty.
+let mut net = NetBuilder::new(0..10)
+    .num_faulty(3)
+    .using(move |node| { DynamicHoneyBadger::builder().build(node.netinfo) })
+    .build()
+    .expect("could not construct test network");
+```
+
+Algorithms that return a `Step` upon construction should use `using_step` instead.
+
+### Sending input
+
+`Input` can be sent to any node of the `VirtualNet` using the `send_input` method:
+
+```rust
+let input = ...;
+let _step = net.send_input(123, input).expect("algorithm failed");
+```
+
+While the resulting step is returned, it needn't be processed to keep the network going, as its messages are automatically added to the queue.
+
+Instead of targeting a node in particular, the same input can be sent to all nodes:
+
+```rust
+net.broadcast_input(input).expect("algorithm failed");
+```
+
+### Cranking the network
+
+The network advances through the `crank()` function, on every call
+
+1. the adversary is given a chance to re-order the message queue,
+1. the next message in the queue is delivered to its destination node (if the node is non-faulty) or the adversary (if the node is faulty),
+1. all messages from the resulting step are queued again,
+1. and the resulting step (or error) is returned.
+
+If there were no messages to begin with, `None` is returned instead.
+
+Cranking can be done manually:
+
+```rust
+let step = net.crank()
+              .expect("expected at least one messages")
+              .expect("algorithm error");
+
+// Shorthand:
+let step = net.crank_expect();
+```
+
+For convenience, an iterator interface is also available:
+
+```rust
+for res in net {
+    let (node_id, step) = res.expect("algorithm error");
+    // ...
+}
+```
+
+This has the drawback that access to the network is not available between cranks, since it will be borrowed inside the for-loop. A common workaround is using a while loop instead:
+
+```rust
+while let Some(res) = net.crank() {
+    let (node_id, step) = res.expect("algorithm error");
+    // `net` can still be mutable borrowed here.
+}
+```
+
+### Inspecting the network
+
+In addition to the returned `Step`s, the network and nodes can be queried through various methods: `VirtualNet::{nodes, faulty_nodes, correct_nodes, get, get_mut}`.
+
+### Adversaries
+
+Adversaries can be introduced through the `.adversary` method on the constructor and are expected to implement the `net::adversary::Adversay` trait. Generic adversaries are available in the same module, while algorithm-specific ones should live next to each test case.
+
+```rust
+// Missing example.
+```
+
+### Tracing
+
+By default, all network tests write traces of every network message into logfiles, named `net-trace_*.txt`. Each log stores one message per line, in the format of `[SENDER] -> [RECEIVER]: MSG`.
```

Comment:
Could you put the path/relative-path to where the log files are written?

Reply:
If I only knew that myself ;). They are opened with no path prefix, so they end up in `cargo test`'s CWD. I'll add an explanation.